### PR TITLE
Add extensions for AndroidX ViewModel with SavedStateHandle factories for #374

### PR DESCRIPTION
**Pull request by carltonwhitehead**
Adds extensions to conveniently obtain ViewModel types from Kodein DI providers/factories with minimal boilerplate for #374.

Example plain ViewModel:
```kotlin
// binding
bindProvider { MyViewModel(
    myDependency = instance()
) }

// retrieval
class MyActivity : AppCompatActivity, DIAware {
    override val di: DI by closestDI()
    private val viewmodel: MyViewModel by viewModel()
} 
```

Example ViewModel with SavedStateHandle:
```kotlin
// binding
bindFactory { savedStateHandle: SavedStateHandle -> MyViewModel(
    savedStateHandle = savedStateHandle,
    myDependency = instance()
) }

// retrieval
class MyActivity : AppCompatActivity, DIAware {
    override val di: DI by closestDI()
    private val viewmodel: MyViewModel by viewModelWithSavedStateHandle()
}
```

### CODE REVIEW

File: framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+dependencies {
+    api(project(":framework:android:kodein-di-framework-android-x"))
+    implementation("androidx.core:core-ktx:1.6.0")
+    implementation("androidx.appcompat:appcompat:1.3.0")
+    implementation("androidx.activity:activity-ktx:1.2.3")
+    implementation("androidx.fragment:fragment-ktx:1.3.5")
+    val lifecycleVersion = "2.3.0"
+    implementation("androidx.lifecycle:lifecycle-viewmodel-ktx:$lifecycleVersion")
+    implementation("androidx.lifecycle:lifecycle-viewmodel-savedstate:$lifecycleVersion")
+    implementation("androidx.lifecycle:lifecycle-livedata-ktx:$lifecycleVersion")
```
Comment (romainbsl) [inline framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts:14]: I'm not sure that this dependency is needed

File: framework/android/kodein-di-framework-android-x/build.gradle.kts [hunk 2]
```diff
@@ -5,7 +5,7 @@ plugins {
 dependencies {
     api(project(":framework:android:kodein-di-framework-android-core"))
 
-    implementation("androidx.appcompat:appcompat:1.0.2")
+    implementation("androidx.appcompat:appcompat:1.3.0")
```
Comment (romainbsl) [inline framework/android/kodein-di-framework-android-x/build.gradle.kts:8]: Latest is `1.3.1`

File: settings.gradle.kts [hunk 3]
```diff
@@ -25,7 +25,8 @@ include(
 android.include(
         ":framework:android:kodein-di-framework-android-core",
         ":framework:android:kodein-di-framework-android-support",
-        ":framework:android:kodein-di-framework-android-x"
+        ":framework:android:kodein-di-framework-android-x",
+        ":framework:android:kodein-di-framework-android-x-viewmodel-savedstate"
```
Comment (romainbsl) [inline settings.gradle.kts:29]: I would suggest to rename the module without the `savedstate` mention, like in the manifest `android.x.viewmodel`.

File: framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+plugins {
+    id("org.kodein.library.android")
+}
+
+dependencies {
+    api(project(":framework:android:kodein-di-framework-android-x"))
+    implementation("androidx.core:core-ktx:1.6.0")
+    implementation("androidx.appcompat:appcompat:1.3.0")
```
Comment (romainbsl) [inline framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts:8]: those aren't the latest version, I suggest to upgrade them to:
```kotlin
    implementation("androidx.appcompat:appcompat:1.3.1")
    implementation("androidx.fragment:fragment-ktx:1.3.6")
```

File: framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    implementation("androidx.activity:activity-ktx:1.2.3")
```
Comment (romainbsl) [inline framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts:9]: this one seems out of scope.

File: framework/android/kodein-di-framework-android-x-viewmodel-savedstate/src/main/java/org/kodein/di/android/x/viewmodel/DIAware.kt [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+ * ```
+ * @param A Extension function applies to [AppCompatActivity] which implements [DIAware]
+ * @param VM The type of [ViewModel] to obtain
+ * @param tag The bound Kodein DI tag
+ * @throws IllegalArgumentException if accessed before the Activity is attached to the Application
+ * @throws DI.NotFoundException If no factory was found.
+ * @throws DI.DependencyLoopException When calling the factory, if the value construction triggered a dependency loop.
+ */
+@MainThread
+inline fun <A, reified VM> A.viewModelSavedStateFactory(
```
Comment (romainbsl) [inline framework/android/kodein-di-framework-android-x-viewmodel-savedstate/src/main/java/org/kodein/di/android/x/viewmodel/DIAware.kt:33]: Can you rename this as `viewModelSavedState` ? 
As the factory is an internal thing et it just returns a lazy viewModel.

File: framework/android/kodein-di-framework-android-x-viewmodel-savedstate/src/main/java/org/kodein/di/android/x/viewmodel/DIAware.kt [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+ * @param F Extension function applies to [androidx.fragment.app.Fragment] which implements [org.kodein.di.DIAware]
+ * @param VM The type of [androidx.lifecycle.ViewModel] to obtain
+ * @param ownerProducer Optionally override the default ViewModel owner
+ * @param tag The bound Kodein DI tag
+ * @throws IllegalArgumentException if accessed before this Fragment is attached
+ * @throws DI.NotFoundException If no factory was found.
+ * @throws DI.DependencyLoopException When calling the factory, if the value construction triggered a dependency loop.
+ */
+@MainThread
+inline fun <F, reified VM> F.viewModelSavedStateFactory(
```
Comment (romainbsl) [inline framework/android/kodein-di-framework-android-x-viewmodel-savedstate/src/main/java/org/kodein/di/android/x/viewmodel/DIAware.kt:72]: Can you rename this as `viewModelSavedState` ? 
As the factory is an internal thing et it just returns a lazy viewModel.

**Review by romainbsl — CHANGES_REQUESTED** [review]
Thank you for ver much your contribution!

Can we add the same extension without the `SavedStateHandle`, giving us something like `by viewModel()` ?

What do you think about adding binding extensions like `bindViewModel` to avoid confusion among users about what to use (singleton / factory / provider / etc):

```kotlin
bindViewModel { savedStateHandle: SavedStateHandle -> MyViewModel(
    savedStateHandle = savedStateHandle,
    myDependency = instance()
) }
```

and 

```kotlin
bindViewModel { MyViewModel(myDependency = instance()) }
```

File: settings.gradle.kts (hunk 3 above)
Comment (carltonwhitehead) [reply settings.gradle.kts:29]: This has an impact on the transitive dependencies that this module would bring into the dependency graph. Katana had separated their basic ViewModel module from their ViewModel with SavedStateHandle module seemingly for this reason. Google packages ViewModel separately from the SavedStateHandle extensions, and apps may well make use of ViewModel without SavedStateHandle extensions. I assumed that Kodein DI would similarly try to minimize the inclusion of potentially unwanted transitive dependencies.

I don't have any too strong feelings about this one way or the other. Proguard/R8 could strip out truly unused components. I would be happy to combine into a single module if that is indeed your preference. Just wanted to raise the potential concern here.

File: framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts (hunk 5 above)
Comment (carltonwhitehead) [reply framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts:9]: I will review and touch this up or report back if there is an issue.

File: framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts (hunk 1 above)
Comment (carltonwhitehead) [reply framework/android/kodein-di-framework-android-x-viewmodel-savedstate/build.gradle.kts:14]: I will review this and tidy up or report back if there is an issue.

**Comment by carltonwhitehead** [comment]
> Thank you for ver much your contribution!
> 
> Can we add the same extension without the `SavedStateHandle`, giving us something like `by viewModel()` ?
> 
> What do you think about adding binding extensions like `bindViewModel` to avoid confusion among users about what to use (singleton / factory / provider / etc):
> 
> ```kotlin
> bindViewModel { savedStateHandle: SavedStateHandle -> MyViewModel(
>     savedStateHandle = savedStateHandle,
>     myDependency = instance()
> ) }
> ```
> 
> and
> 
> ```kotlin
> bindViewModel { MyViewModel(myDependency = instance()) }
> ```

I like the idea of the binding extensions. I will also add factory extensions for plain ViewModels.

File: settings.gradle.kts (hunk 3 above)
Comment (romainbsl) [reply settings.gradle.kts:29]: Ok then, let's go for the module `:framework:android:kodein-di-framework-android-x-viewmodel-savedstate` to follow Google practices.

So, for the plain ViewModel you will go with another module `:framework:android:kodein-di-framework-android-x-viewmodel` ?

Comment (carltonwhitehead) [reply settings.gradle.kts:29]: Sounds good. I will do separate modules with those names.

**Comment by carltonwhitehead** [comment]
Naming the the extensions for both plain provider and SavedStateHandle factory styles with the same name is problematic because it creates overload ambiguity errors. I think we will have to name them differently to avoid it. I propose:

```kotlin
// plain ViewModel extensions
bindViewModel { MyViewModel(
  myDep = instance()
) }
bind { scoped(MyScope).viewModel { ScopedViewModel(
  scope = context, 
  myDep = instance()
) } }

// SavedStateHandle ViewModel extensions
bindViewModelWithSavedStateHandle { MySavedStateViewModel(
  state = it, 
  myDep = instance()
) }
bind { scoped(MyScope).viewModelWithSavedStateHandle { ScopedViewModel(
  scope = context, 
  state = it, 
  myDep = instance()
) } }
```

I will note this doesn't save on boilerplate compared to using the standard binding style.

```kotlin
bind { scoped(MyScope).factory { savedStateHandle: SavedStateHandle(
  scope = context,
  state = it,
  myDep = instance()
) } }
```

Please advise.

**Comment by romainbsl** [comment]
Indeed, the idea of having the same name was bad, we can go for both `bindViewModel` and `bindViewModelWithSavedStateHandle` that matches most of use cases iMHO. 

For the  scope dependencies I guess we can stand with the standard binding functions of Kodein-DI, and document it properly.

**Comment by carltonwhitehead** [comment]
> Indeed, the idea of having the same name was bad, we can go for both `bindViewModel` and `bindViewModelWithSavedStateHandle` that matches most of use cases iMHO.
> 
> For the scope dependencies I guess we can stand with the standard binding functions of Kodein-DI, and document it properly.
> 
> What do you think of it ?

I've changed my mind about ViewModel-specific binding extensions. I don't think they add value. Internally, they just call down to corresponding Provider or Factory. They don't save the developer much, if any, boilerplate code compared to using plain binding style. The developer still needs to know they're writing a provider or factory with SavedStateHandle argument. It feels cleaner to me to just use Kotlin language features to accomplish that.

**Comment by romainbsl** [comment]
Fine by me, we will document this in a proper way.

**Comment by carltonwhitehead** [comment]
Just pushed a batch of commits which I believe address all of the changes requested so far.

**Review by romainbsl — APPROVED** [review]
Thank you so much for your contribution 🙏 
Can you had a section in the documentation about this in 
https://github.com/Kodein-Framework/Kodein-DI/blob/master/doc/modules/framework/pages/android.adoc ?

Or do you want me to do it?

**Comment by carltonwhitehead** [comment]
I will add some documentation.

**Review by romainbsl — APPROVED** [review]
